### PR TITLE
Re-implement 3p license inference implementation.

### PR DESCRIPTION
**Pull request by vkryachko**
The implementation relied on an oss gradle plugin. The plugin had
several issues:

* Did not respect project level dependencies
* Did not respect some gradle configurations that propagate to the pom,
  like runtimeOnly
* Had resolution issues where the order of task execution would affect
  the outcome of the license report
* Failed the build for projects that have debug buildType disabled.

### CODE REVIEW

File: buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/GenerateLicensesTask.groovy [hunk 1]
```diff
@@ -15,37 +15,54 @@
...
 
+    @Inject
+    GenerateLicensesTask(Configuration configuration) {
+        this(configuration, new LicenseResolver())
+    }
+
+    GenerateLicensesTask(Configuration configuration, LicenseResolver licenseResolver) {
+        this.inputConfiguration = configuration
+        this.licenseResolver = licenseResolver
+        outputs.upToDateWhen { false }
```
Comment (ashwinraghav) [inline buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/GenerateLicensesTask.groovy]: We always run the task? Could you add a comment why?

File: buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/LicenseResolver.groovy [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            Project depProject = ((ProjectDependency) it).dependencyProject
+            Set<ProjectLicense> transitive = determineProjectLicenses(depProject.configurations.getByName(configuration.name))
+            return [determineLicense(depProject)] + transitive
+        }.findAll { it != null } as Set
+    }
+
+    private static Set<ProjectLicense> determineExternalLicenses(Project project, Configuration configuration) {
+        def conf = project.configurations.create('internalAllExternalArtifacts')
+
+        configuration.allDependencies.each {
```
Comment (ashwinraghav) [inline buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/LicenseResolver.groovy]: Four our future selves, can you leave a comment on why these tactics are necessary?

File: buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/ProjectLicense.groovy [hunk 3]
```diff
@@ -12,10 +12,69 @@
...
+            def name = project.name
+            def licenses = project.licenses.license.collect { URI.create(it.url as String) }
+            return new ProjectLicense(
+                    name: name,
+                    explicitLicenseUris: licenses,
+                    parent: determineParent(pomFile.parentFile.parentFile.parentFile.parentFile.parentFile, project.parent))
+        }
+    }
+
+    private static ProjectLicense determineParent(File baseDir, GPathResult parentNode) {
```
Comment (ashwinraghav) [inline buildSrc/src/main/groovy/com/google/firebase/gradle/plugins/license/ProjectLicense.groovy]: Worth leaving a comment on why we need to look up licenses from the parent for some libs

File: buildSrc/src/test/groovy/com/google/firebase/gradle/plugins/license/LicenseResolverPluginSpec.groovy [hunk 4]
```diff
@@ -80,7 +80,7 @@ class LicenseResolverPluginSpec extends Specification {
```
Comment (ashwinraghav) [inline buildSrc/src/test/groovy/com/google/firebase/gradle/plugins/license/LicenseResolverPluginSpec.groovy]: Wow! So the test actually served a purpose and passess?

Comment (vkryachko) [reply buildSrc/src/test/groovy/com/google/firebase/gradle/plugins/license/LicenseResolverPluginSpec.groovy]: yep, with one minor capitalization change

**Pull request merged**